### ISSUE TITLE
Compiling and launching the game from Docker container

### ISSUE THREAD

**Issue by randomnoise** [OpenRA/OpenRA#20997]
First of all, thanks for the project.

### Motivation
As a person who is interested in OpenRA and a newcomer to the repo, I wanted to the see the game's working condition from `bleed` branch. I started by reading the `README.md` and some linked files, clone the repo to my Linux machine, compile the code.

The code compiled fine, but when I tried to launch the game, it didn't work. I've tried some possible solutions but I could not get it to work (more details down below).

This creates a chilling effect for newcomers who is interested in but cannot start to contribute to the project without working code execution.

### Proposed solution
Inspired by Jess Frazelle's somewhat old [blog post](https://blog.jessfraz.com/post/docker-containers-on-the-desktop/) about using desktop GUI apps working in `Docker` containers, I feel the project should include files for `Docker image`s to launch the game **inside containers** and/or compile the code.

`Dockerfile`s compel developers to declare the underlying dependencies explicitly, they create a portable images between different distros and Docker kind of solves the "works on other machines" problems.

I have created `Docker` files and shell script in working condition. I will send a PR recently, but I also wanted to open an issue for a more general discussion about using containers in the project.

Whether the proposal will be accepted, Docker image has allowed me to start to contribute to the repo, so I am guessing I will continue to use the `Docker image` even if I cannot solve my initial problem. I hope, by creating a searchable artifact in the repo, others in the community will benefit in some way hopefully.

### Side effects
Working with GUI apps working in containers is solved problem for Linux; but for MacOS and Windows, it may not work as well. I have no idea about Windows; in MacOS, there are some approaches that are kinda slow.

So, the Docker solution is related to working with `dotnet` and `mono` artifacts in Linux. The game working inside the Docker container works  performantly in Linux.

<details>

<summary>Tried alternatives (TL;DR - problem is the snap package)</summary>

### Alternatives: A journey

I wanted to give the games a spin, I've tried `release` and `playtest-*` AppImages, they worked flawlessly.

I wanted to get involved a bit, test the `bleed` branch's games; followed the instructions in Markdown files, install required dependencies, compiled the code, so far so good; but when I tried the launch a game from the compiled code, it did not work. The message said "the renderer initialization failed". 

I wanted to solve this initial problem and I wanted try some possible avenues before writing something publicly; if AppImages worked without a problem, I thought, how big of an issue can it be? Boy, was I wrong... The rest is a kind of a rabbit hole and a bit of a journey (I also wanted to document this part for future reference and for sake of others):

* Looked around [FAQs](https://github.com/OpenRA/OpenRA/wiki/FAQ) and old issues, followed threads to other repos but I couldn't find something concrete.

Looking around AppImages:
* Tried tracing systems calls of AppImages, I thought there might be undocumented, implicit  dependencies that I was missing. I think because they are published in a [--self-contained](https://github.com/OpenRA/OpenRA/blob/2744b44d93/packaging/functions.sh#L83) way, they don't need to make so many system calls relatively (also my search with regex and `strace` skills are limited, it's possible I might have missed something).

* Looked for the latest GitHub Action [logs](https://github.com/OpenRA/OpenRA/actions/workflows/packaging.yml) for `playtest` Linux AppImage build; looks like other tools create some noise in the logs, but not much entry related to OpenRA, I could not find useful info in the logs towards solving my issue. (I understand it is mostly unnecessary info for a veteran, but as a newbie I want everything to be verbose 🥲)

* Wanted to extract info and dependencies from the AppImage file itself, I did not work with the format before, I could not do it.

Looking around files that compiled locally:
* Tried working with `make TARGETPLATFORM=linux-x64` and `make TARGETPLATFORM=unix-generic` commands. They did not create working solution.

* For both `TargetPlatform`s, neither console outputs, `graphics.log` or `exception.log`s gave enough explanatory clues to work towards a solution.

* `strace`d for `TargetPlatform=linux-x64`: app wanted to connect to a non-existent `wayland-0` socket, I could not solve that.

* `strace`d for `TargetPlatform=unix-generic`: it is weird, could find some SDL2 object files, could not find others, I've tried to work with newer version from [the SDL website](https://www.libsdl.org/), tried [linking them properly](https://github.com/OpenRA/OpenRA/blob/2744b44d93/configure-system-libraries.sh#L73), but to no avail.

In the end, I think the source of the problem is a configuration issue on my local machine, but could not the proper solution to that. Also, I use dotnet-sdk from `snap`, that may play a part.

</details>

### Conclusion

It is bit of a read, it can even be a blog post with some screenshots and more granular details but I think it needs to documented here for search.

I am new to project and to the community, so any feedback is welcome 😊

**Comment by Mailaender** [comment]
Can you explain why you couldn't run the game after compiling? Putting it into containers only adds complexity. I think you should have asked for help in https://www.openra.net/community/ places first.

**Comment by Mailaender** [comment]
> Also, I use dotnet-sdk from `snap`, that may play a part.

I am not that experienced with snaps, but I think their .NET SDK is designed to be used with other snaps only. https://learn.microsoft.com/dotnet/core/install/linux is the way to go.

**Comment by randomnoise** [comment]
Thanks for the quick reply and sorry for the somewhat confusing and long post there (I made the exploration part collapsible to make it more bearable in first post)

Snaps are supported by Microsoft and they provide [snap installation links](https://learn.microsoft.com/en-us/dotnet/core/install/linux-snap) themselves. I preferred them for the newer versions (snap offer dotnet-sdk version 6.0.412, apt offers version 6.0.120)

**But,** it turns out the problem was the snap package. I removed that, tried the apt package, worked like a charm. Deep exploration in the original post was a bit unnecessary but educational 😀

Although, my proposal about development Docker images still stands whether it's accepted. To some, using containers like that feels like an abomination, and I get that argument.
Also, some Linux distros does not work well with `dotnet-sdk`, Microsoft does not offer info about packages about some major rolling distros. Dockerfiles are expressive about explicit dependencies and for that, I find them useful.
